sympy/ntheory/qs.py

```diff
@@ -0,0 +1,234 @@
+from sympy import isprime, log, sqrt, exp, sieve, gcd, mod_inverse, ceiling
+from sympy.ntheory.residue_ntheory import _sqrt_mod_prime_power
+
+class Polynomial:
+    def __init__(self, modified_coeff=[], a=None, b=None):
+        """ Here a and b are the parameters of the polynomial, and
+        modified_coeff is modified coeff to reduce amount of
+        calculation whenever eval is called.
+        modified_coeff = [a**2, 2*a*b, b**2 - N]"""
+        self.modified_coeff = modified_coeff
+        self.a = a
+        self.b = b
+
+    def eval(self, x):
+        ans = 0
+        for coeff in self.modified_coeff:
+            ans *= x
+            ans += coeff
+        return ans
+
+class factor_base_elem:
+    def __init__(self, prime, tmem_p, log_p):
+        self.prime = prime
+        self.tmem_p = tmem_p
+        self.log_p = log_p
+        self.soln1 = None
+        self.soln2 = None
+        self.a_inv = None
+        self.B_ainv = None
+
+
+
+def gen_factor_base(B, n):
+    factor_base = [2, 3]
+    for p in sieve.primerange(4, B + 1):
+        a = n % p
+        if a < 2 or pow(a, (p - 1) // 2, p) == 1:
+            residue = _sqrt_mod_prime_power(n, p, 1)
+            log_p = round(log(p, 2).evalf())
```

Comment:
It is not explained in the source what "rounding off" should mean but this may not be the best implementation because most digits of the logarithm are lost. The algorithm would be valid with floating point numbers except that addition would be slightly less efficient. I think that rounding here should preserve a number of decimals after the binary point like

    round(log(p, 2).evalf(), 10)

The result will be a float representing a "fixed point" number. I would convert it to an integer in order to speed up the addition by something like

    round(log(p, 2).evalf()*2**30)

Finally, it is not necessary to use base 2 logarithms, any base would work for the algorithm. Computation of the natural logarithm `log(p)` might be most efficient, so I would suggest something like

    round(log(p).evalf()*2**30)

Reply:
Any reason not to just write `round(log(p)*2**30)`?

Reply:
It seems to work as well because of `round`, but apparently not without it:
```
In [27]: log(37)*2**30
Out[27]: 1073741824⋅log(37)

In [28]: round(log(37)*2**30)
Out[28]: 3877193586
```